env/ipp_aes_ctr_provider.cc

```diff
@@ -0,0 +1,204 @@
+                                          size_t prefixLength) const {
+#if defined(IPPCP) && (defined(HAVE_SSE42) || defined(HAVE_SSE2))
+  IppStatus ipp_status;
+  Ipp32u rnd;
+  const size_t rnd_size = sizeof(Ipp32u);
+  assert(prefixLength % rnd_size == 0);
+  for (size_t i = 0; i < prefixLength; i += rnd_size) {
+    ipp_status = ippsPRNGenRDRAND(&rnd, rnd_size << 3, nullptr);
+    if (ipp_status != ippStsNoErr)
+      return Status::Aborted(ippcpGetStatusString(ipp_status));
+    memcpy(prefix + i, &rnd, rnd_size);
+  }
+  IppAESCTRCipherStream cs(aes_ctx_, prefix);
+  return cs.Encrypt(0, prefix + IppAESCTRCipherStream::kBlockSize,
+                    prefixLength - IppAESCTRCipherStream::kBlockSize);
+#else
+  (void)prefix;
+  (void)prefixLength;
+  return Status::OK();
+#endif  // IPPCP
+}
+
+Status IppAESCTRProvider::CreateCipherStream(
+    const std::string& /*fname*/, const EnvOptions& /*options*/, Slice& prefix,
+    std::unique_ptr<BlockAccessCipherStream>* result) {
+#if defined(IPPCP) && (defined(HAVE_SSE42) || defined(HAVE_SSE2))
+  assert(result != nullptr);
+  assert(prefix.size() >= IppAESCTRCipherStream::kBlockSize);
+  result->reset(new IppAESCTRCipherStream(aes_ctx_, prefix.data()));
+  Status ipp_status = (*result)->Decrypt(
+      0, (char*)prefix.data() + IppAESCTRCipherStream::kBlockSize,
+      prefix.size() - IppAESCTRCipherStream::kBlockSize);
+  return ipp_status;
+#else
+  (void)prefix;
+  (void)result;
+  return Status::OK();
+#endif  // IPPCP
+}
+
+IppAESCTRProvider::~IppAESCTRProvider() {
+#if defined(IPPCP) && (defined(HAVE_SSE42) || defined(HAVE_SSE2))
+  ippsAESInit(0, static_cast<int>(key_size_), aes_ctx_, ctx_size_);
+  delete[](Ipp8u*) aes_ctx_;
+#endif  // IPPCP
+}
+
+#if defined(IPPCP) && (defined(HAVE_SSE42) || defined(HAVE_SSE2))
+
+IppAESCTRCipherStream::IppAESCTRCipherStream(IppsAESSpec* aes_ctx,
+                                             const char* init_vector)
+    : aes_ctx_(aes_ctx) {
+  init_vector_ = _mm_loadu_si128((__m128i*)init_vector);
+}
+
+Status IppAESCTRCipherStream::Encrypt(uint64_t fileOffset, char* data,
+                                      size_t dataSize) {
+  if (dataSize == 0) return Status::OK();
+
+  size_t index = fileOffset / kBlockSize;
+  size_t offset = fileOffset % kBlockSize;
+
+  Ipp8u ctr_block[kBlockSize];
+
+  __m128i counter = _mm_add_epi64(init_vector_, _mm_cvtsi64_si128(index));
+  Ipp8u* ptr_counter = (Ipp8u*)&counter;
+  for (size_t i = 0; i < kBlockSize; ++i)
+    ctr_block[i] = ptr_counter[kBlockSize - 1 - i];
+
+  IppStatus ipp_status = ippStsNoErr;
+  if (offset == 0) {
+    ipp_status = ippsAESEncryptCTR((Ipp8u*)(data), (Ipp8u*)data, dataSize,
+                                   aes_ctx_, ctr_block, kCounterLen);
+  } else {
+    Ipp8u zero_block[kBlockSize]{0};
+    ipp_status = ippsAESEncryptCTR(zero_block, zero_block, kBlockSize, aes_ctx_,
+                                   ctr_block, kCounterLen);
```

Comment:
This status is never checked and overwritten a few lines later.

Reply:
Fixed.